comid/group.go

```diff
@@ -75,43 +60,97 @@ func (o *Group) UnmarshalCBOR(data []byte) error {
 //	  "value": "69E027B2-7157-4758-BCB4-D9F167FE49EA"
 //	}
 func (o *Group) UnmarshalJSON(data []byte) error {
-	var v tnv
+	var tnv encoding.TypeAndValue
 
-	if err := json.Unmarshal(data, &v); err != nil {
+	if err := json.Unmarshal(data, &tnv); err != nil {
+		return fmt.Errorf("group decoding failure: %w", err)
+	}
+
+	decoded, err := NewGroup(nil, tnv.Type)
+	if err != nil {
 		return err
 	}
 
-	switch v.Type {
-	case "uuid":
-		var x UUID
-		if err := x.UnmarshalJSON(v.Value); err != nil {
-			return err
-		}
-		o.val = TaggedUUID(x)
-	default:
-		return fmt.Errorf("unknown type %s for group", v.Type)
+	if err := json.Unmarshal(tnv.Value, &decoded.Value); err != nil {
+		return fmt.Errorf(
+			"cannot unmarshal group: %w",
+			err,
+		)
 	}
 
+	if err := decoded.Value.Valid(); err != nil {
+		return fmt.Errorf("invalid %s: %w", tnv.Type, err)
+	}
+
+	o.Value = decoded.Value
+
 	return nil
 }
 
 func (o Group) MarshalJSON() ([]byte, error) {
-	var (
-		v   tnv
-		b   []byte
-		err error
-	)
-
-	switch t := o.val.(type) {
-	case TaggedUUID:
-		b, err = UUID(t).MarshalJSON()
-		if err != nil {
-			return nil, err
-		}
-		v = tnv{Type: "ueid", Value: b}
-	default:
-		return nil, fmt.Errorf("unknown type %T for group", t)
+	return extensions.TypeChoiceValueMarshalJSON(o.Value)
+}
+
+type IGroupValue interface {
+	extensions.ITypeChoiceValue
+}
+
+func NewUUIDGroup(val any) (*Group, error) {
+	if val == nil {
+		return &Group{&TaggedUUID{}}, nil
+	}
+
+	u, err := NewTaggedUUID(val)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Group{u}, nil
+}
+
+func MustNewUUIDGroup(val any) *Group {
+	ret, err := NewUUIDGroup(val)
+	if err != nil {
+		panic(err)
+	}
+
+	return ret
+}
+
+// IGroupFactory defines the signature for the factory functions that may be
+// registred using RegisterGroupType to provide a new implementation of the
+// corresponding type choice. The factory function should created a new *Group
```

Comment:
a pattern starts emerging :)

```suggestion
// corresponding type choice. The factory function should create a new *Group
```